16-encoding-categorical-data.Rmd

```diff
@@ -1,8 +1,356 @@
+
+:::rmdnote
+We advise starting with untransformed categorical variables when a model allows it, and point out that more complex encodings often do not result in better performance for such models.
+:::
+
+## Using the outcome for encoding predictors
+
+There are multiple options for encodings more complex than dummy or indicator variables. One method called _effect_ or _likelihood encodings_ replaces the original categorical variables with a single numeric column that measures the effect of those data. For example, for the neighborhood predictor in the Ames housing data, we can compute the mean or median sale price for each neighborhood and substitute these means for the original data values. 
+
+```{r}
+ggplot(ames_train, 
+       aes(y = reorder(Neighborhood, Sale_Price), x = Sale_Price)) + 
+  geom_bar(stat = "summary", fun = "mean") + 
+  labs(y = NULL)
+```
+
+This kind of effect encoding works well when your categorical variable has many levels. In tidymodels, the `r pkg(embed)` package includes several recipe step functions for different kinds of effect encodings, such as `step_lencode_glm()`, `step_lencode_mixed()`, and `step_lencode_bayes()`. These steps use a generalized linear model to estimate the effect of each level in a categorical predictor on the outcome.
+
+
+```{r}
+library(embed)
+
+ames_glm <- 
+  recipe(Sale_Price ~ Neighborhood + Gr_Liv_Area + Year_Built + Bldg_Type + 
+           Latitude + Longitude, data = ames_train) %>%
+  step_log(Gr_Liv_Area, base = 10) %>% 
+  step_lencode_glm(Neighborhood, outcome = vars(Sale_Price)) %>%
+  step_dummy(all_nominal_predictors()) %>% 
+  step_interact( ~ Gr_Liv_Area:starts_with("Bldg_Type_") ) %>% 
+  step_ns(Latitude, Longitude, deg_free = 20)
+
+ames_glm
+```
+
+:::rmdnote
+When using a recipe step like `step_lencode_glm()`, specify the variable being encoded first and then the outcome using `vars()`.
+:::
+
+As detailed in Section \@ref(recipe-functions), we can `prep()` our recipe to fit or estimate parameters for the preprocessing transformations using training data. We can then `tidy()` this prepared recipe to see the results.
+
+```{r}
+glm_estimates <-
+  prep(ames_glm) %>%
+  tidy(number = 2)
+
+glm_estimates
+```
+
+When we use the newly encoded `Neighborhood` numeric variable created via this method, we substitute the original level (such as `"North_Ames"`) with the estimate for `Sale_Price` from the GLM.
+
+Effect encoding methods like this one can also seamlessly handle situations where a novel factor level is encountered in the data. 
+
+```{r}
+glm_estimates %>%
+  filter(level == "..new")
+```
+
+This is the predicted price from the GLM when we don't have any specific neighborhood information.
+
+:::rmdwarn
+Effect encodings can be powerful but should be used with care.
```

Comment:
```suggestion
Effect encodings can be powerful but should be used with care. The effects should be computed from the training set (and not prior to data splitting or modeling). Also, this type of _supervised_ preprocessing should be thoroughly resampled to avoid overfitting. 
```